--- a/CRM/Utils/System/Drupal.php
+++ b/CRM/Utils/System/Drupal.php
@@ -304,6 +304,38 @@ class CRM_Utils_System_Drupal extends CRM_Utils_System_DrupalBase {
   }

   /**
+   * Check if a resource url is within the drupal directory and format appropriately
+   *
+   * @param url (reference)
+   *
+   * @return bool: TRUE for internal paths, FALSE for external. The drupal_add_js fn is able to add js more
+   * efficiently if it is known to be in the drupal site
+   */
+  function formatResourceUrl(&$url) {
+    $internal = FALSE;
+    $base = CRM_Core_Config::singleton()->resourceBase;
+    global $base_url;
+    // Handle absolute urls
+    // compares $url (which is some unknown/untrusted value from a third-party dev) to the CMS's base url (which is independent of civi's url)
+    // to see if the url is within our drupal dir, if it is we are able to treated it as an internal url
+    if (strpos($url, $base_url) === 0) {
+      $internal = TRUE;
+      $url = trim(str_replace($base_url, '', $url), '/');
+    }
+    // Handle relative urls that are within the CiviCRM module directory
+    elseif (strpos($url, $base) === 0) {
+      $internal = TRUE;
+      $url = $this->appendCoreDirectoryToResourceBase(substr(drupal_get_path('module', 'civicrm'), 0, -6)) . trim(substr($url, strlen($base)), '/');
+    }
+    // Strip query string
+    $q = strpos($url, '?');
+    if ($q && $internal) {
+      $url = substr($url, 0, $q);
+    }
+    return $internal;
+  }
+
+  /**
    * @inheritDoc
    */
   public function authenticate($name, $password, $loadCMSBootstrap = FALSE, $realPath = NULL) {
@@ -469,7 +501,9 @@ AND    u.status = 1
     }
     // load drupal bootstrap
     chdir($cmsPath);
-    define('DRUPAL_ROOT', $cmsPath);
+    if(!defined('DRUPAL_ROOT')) {
+      define('DRUPAL_ROOT', $cmsPath);
+    }

     // For drupal multi-site CRM-11313
     if ($realPath && strpos($realPath, 'sites/all/modules/') === FALSE) {
@@ -635,11 +669,11 @@ AND    u.status = 1
       return $url;
     }

-    //CRM-7803 -from d7 onward.
+    ///CRM-7803 -from d7 onward.
     $config = CRM_Core_Config::singleton();
-    if (function_exists('variable_get') &&
-      module_exists('locale') &&
-      function_exists('language_negotiation_get')
+    if (function_exists('variable_get') &&
+      function_exists('language_negotiation_get') &&
+      module_exists('locale')
     ) {
       global $language;

