--- a/profiles/civicrm_starterkit/modules/civicrm/CRM/Contribute/Form/Task/PDFLetterCommon.php
+++ b/profiles/civicrm_starterkit/modules/civicrm/CRM/Contribute/Form/Task/PDFLetterCommon.php
@@ -325,8 +325,18 @@ public static function buildContributionArray($groupBy, $contributionIDs, $retur
    */
   public static function combineContributions($existing, $contribution, $separator) {
     foreach ($contribution as $field => $value) {
-      $existing[$field] = isset($existing[$field]) ? $existing[$field] . $separator : '';
-      $existing[$field] .= $value;
+      $existingValue = '';
+      if (isset($existing[$field])) {
+        $existingValue = $existing[$field];
+        if ($existingValue == '') {
+         $existingValue .= '&nbsp;';
+        }
+        $existingValue .= $separator;
+      }
+      if ($value == '') {
+        $value = '&nbsp;';
+      }
+      $existing[$field] = $existingValue . $value;
     }
     return $existing;
   }
--- a/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Token.php
+++ b/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Token.php
@@ -1725,7 +1725,12 @@ public static function replaceMultipleContributionTokens($separator, $str, &$con
       $formattedDates = array();
       $dates = explode($separator, $contribution['receive_date']);
       foreach ($dates as $date) {
-        $formattedDates[] = CRM_Utils_Date::customFormat($date, NULL, array('j', 'm', 'Y'));
+        if (strlen($date) == 0) {
+          $formattedDates[] = $date;
+        }
+        else {
+          $formattedDates[] = CRM_Utils_Date::customFormat($date, NULL, array('j', 'm', 'Y'));
+        }
       }
       $str = str_replace("{contribution.receive_date}", implode($separator, $formattedDates), $str);
       unset($knownTokens['contribution']['receive_date']);
