--- a/CRM/Core/BAO/Dashboard.php
+++ b/CRM/Core/BAO/Dashboard.php
@@ -164,7 +164,7 @@ class CRM_Core_BAO_Dashboard extends CRM_Core_DAO_Dashboard {
     $contactID = CRM_Core_Session::singleton()->get('userID');
     $allDashlets = CRM_Utils_Array::index(array('name'), $getDashlets['values']);
     $defaultDashlets = array();
-    $defaults = array('blog' => 1, 'getting-started' => '0');
+    $defaults = array();
     foreach ($defaults as $name => $column) {
       if (!empty($allDashlets[$name])) {
         $defaultDashlets[$name] = array(
