diff --git a/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Check/Component/Security.php b/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Check/Component/Security.php
index 7d763dd4b..89a7131f1 100644
--- a/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Check/Component/Security.php
+++ b/profiles/civicrm_starterkit/modules/civicrm/CRM/Utils/Check/Component/Security.php
@@ -262,17 +262,20 @@ class CRM_Utils_Check_Component_Security extends CRM_Utils_Check_Component {
   public function checkRemoteProfile() {
     $messages = array();
 
-    if (Civi::settings()->get('remote_profile_submissions')) {
-      $messages[] = new CRM_Utils_Check_Message(
-        __FUNCTION__,
-        ts('Warning: External profile support (aka "HTML Snippet" support) is enabled in <a href="%1">system settings</a>. This setting may be prone to abuse. If you must retain it, consider HTTP throttling or other protections.',
-          array(1 => CRM_Utils_System::url('civicrm/admin/setting/misc', 'reset=1'))
-        ),
-        ts('Remote Profiles Enabled'),
-        \Psr\Log\LogLevel::WARNING,
-        'fa-lock'
-      );
-    }
+    // We've commented out this code to disable this message, since we are using
+    // this feature and don't want users to see this message all the time.
+
+    // if (Civi::settings()->get('remote_profile_submissions')) {
+    //   $messages[] = new CRM_Utils_Check_Message(
+    //     __FUNCTION__,
+    //     ts('Warning: External profile support (aka "HTML Snippet" support) is enabled in <a href="%1">system settings</a>. This setting may be prone to abuse. If you must retain it, consider HTTP throttling or other protections.',
+    //       array(1 => CRM_Utils_System::url('civicrm/admin/setting/misc', 'reset=1'))
+    //     ),
+    //     ts('Remote Profiles Enabled'),
+    //     \Psr\Log\LogLevel::WARNING,
+    //     'fa-lock'
+    //   );
+    // }
 
     return $messages;
   }
